orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -869,89 +869,217 @@ common::Status HandleSharedInitializer(Graph& graph,
+  ORT_RETURN_IF_ERROR(
+    AddNewScalarNodeArgAndInitializer<bool>(graph,
+                                            "recv_input_signal" + cut_index_str,
+                                            ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                                            true, /* initializer data */
+                                            recv_input_args,
+                                            new_input_names));
+  ORT_RETURN_IF_ERROR(
+    AddNewScalarNodeArgAndInitializer<size_t>(graph,
+                                              "recv_src_rank" + cut_index_str,
+                                              ONNX_NAMESPACE::TensorProto_DataType_INT64,
+                                              current_stage, /* initializer data */
+                                              recv_input_args,
+                                              new_input_names));
+
+  // add output node_arg for send/recv
+  AddNewNodeArg(graph, "send_output_signal" + cut_index_str,
+                ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                send_output_args, new_output_names);
+
+  AddNewNodeArg(graph, "receive_output_signal" + cut_index_str,
+                ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                recv_output_args, new_output_names);
+
+  return Status::OK();
+}
+
+common::Status SplitGraphWithMap(Graph& graph,
+                                 std::map<Node*, int>& op_to_stage,
+                                 int nstages,
+                                 std::vector<std::pair<int, int>>& messages,
+                                 std::vector<Node*>& send_nodes,
+                                 std::vector<Node*>& recv_nodes) {
+
+  // forward_messages[s]: all the tensors sent by stage s while executing
+  // forward computation.
+  std::vector<std::set<const NodeArg*>> forward_messages(nstages);
+  // TODO(jufranc): once we start using this function on the training graph,
+  // we need to keep backward_messages[s] too.
+
+  // Tensors that need to be sent from one device to the other.
+  // TODO(jufranc): Should we consider weights here too? 
+  // forwarded_tensors[i] = {t, {stage of producer, stage of the last consumer}}
+  std::vector<std::pair<const NodeArg*, std::pair<int, int>>> forwarded_tensors;
+
+  // All the tensors that are produced and consumed in the graph.
+  auto initial_node_args = GetAllNodeArgs(graph);
+
+  // We create all the tensor replicas in advance using this fuction.
+  // A tensor produced in stage r and consumed in stage r', such that r' > r,
+  // will have a replica in all stages r'', such that r'' > r and r'' < r'.
+  // tensor_replicas[t][r] contains a pointer to the the replica of t in stage r, 
+  // it if exists, or to itself if r is the stage of the producer of r.
+  std::map<const NodeArg*, std::vector<NodeArg*>> tensor_replicas;
+  auto create_tensor_replica = [&tensor_replicas, &graph](const NodeArg* tensor,
```

Comment:
>tensor [](start = 73, length = 6)

If the code only supports tensor types, we need to throw if "tensor" is not a tensor.

Reply:
Is there such a function in ORT for that? 

This part of the code is very similar to the [original one](https://github.com/microsoft/onnxruntime/blob/master/orttraining/orttraining/core/graph/pipeline_transformer.cc#L1018).  So, I wonder if maybe I just used a bad name for that variable. Are scalars and constants treated differently in the original version? 

@xzhu1900, is this check missing in the original code too?

Reply:
For training, we only use tensors. Since we are also trying pipeline parallel for inference, it's better to consider non-tensor cases here. I don't think we need to support non-tensor types now, but we can generate a better error message here.

---
In reply to: [537637435](https://github.com/microsoft/onnxruntime/pull/5940#discussion_r537637435) [](ancestors = 537637435)

Reply:
Added the message: "Only tensors are supported in training."